✨ Add card option to more-info action to show a dialog with specified card in it

iantrich opened the pull request:
Allows you to create a dialog with a card in it.

```yaml
type: entity-button
tap_action:
  action: toggle
hold_action:
  action: card
  card:
    type: entity-button
    entity: light.kitchen
double_tap_action:
  action: more-info
show_icon: true
show_name: true
entity: light.kitchen
```
![image](https://user-images.githubusercontent.com/1287159/67454844-4c726680-f5f1-11e9-9f46-a647a3bb07c3.png)
Should probably remove that bottom padding

Closes https://github.com/home-assistant/home-assistant-polymer/issues/3998
Docs: TODO

iantrich committed:
New card action to show a dialog with specified card in it

iantrich commented:
Need to play with this some more; how navigation actions will work, a card action within a card action and more.

Misiu commented:
@iantrich if I understand correctly this will allow us to replace the group with a proper, modern solution.
This way issues [like this](https://github.com/home-assistant/home-assistant-polymer/issues/3998) one can be fixed?

I have two questions:
- can this action be added to a single entity in entities card? (I think yes, but I want to confirm that) I'd like to use one entity from entities card to open a dialog with another entities card
- can we have a close button and title in that dialog? For example, if I want my entity to open cover automation setting, I'd like to be able to specify a title, like "Bedroom window settings"

iantrich commented:
> @iantrich if I understand correctly this will allow us to replace the group with a proper, modern solution.
> This way issues [like this](https://github.com/home-assistant/home-assistant-polymer/issues/3998) one can be fixed?
> 
> I have two questions:
> 
> * can this action be added to a single entity in entities card? (I think yes, but I want to confirm that) I'd like to use one entity from entities card to open a dialog with another entities card
> * can we have a close button and title in that dialog? For example, if I want my entity to open cover automation setting, I'd like to be able to specify a title, like "Bedroom window settings"

1. Yes
2. I'm going to lean towards, no. If you want a title, use a stack

Misiu commented:
> 1. Yes

👍

> 2. I'm going to lean towards, no. If you want a title, use a stack

I've searched over the repo for usages of `ha-paper-dialog` and I found, for example, [this one](https://github.com/home-assistant/home-assistant-polymer/blob/a889a02e156c3fcd5d7649b69454d4a491fe8b44/src/dialogs/domain-toggler/dialog-domain-toggler.ts#L35), it's just a matter of adding `h2` with the title.
If I understand correctly You will be able to specify card in action called 'card` and set all the standard properties that a card has, so this should be a valid config:

```yaml
type: entities
title: Entities card sample
show_header_toggle: false
entities:
  - entity: alarm_control_panel.alarm
    name: Alarm Panel
    hold_action:
      action: card
        card:
          type: entities
          title: Details
          show_header_toggle: false
            entities:
              - switch.a_lights
              - switch.b_lights
  - device_tracker.demo_paulus
  - switch.decorative_lights
  - group.all_lights
  - group.all_locks
```

Please reconsider adding this.

bramkragten commented on src/panels/lovelace/dialogs/card/dialog-card.ts:
```diff
@@ -0,0 +1,85 @@
+  private _renderCard(config: LovelaceCardConfig): TemplateResult {
+    const element = createCardElement(config);
+    if (this.hass) {
+      element.hass = this.hass;
```
Card should also be updated when hass changes

bramkragten commented:
Not sure about this one, it is a cool feature, but a nightmare config wise

iantrich commented:
> Not sure about this one, it is a cool feature, but a nightmare config wise

Don't make me go and make my own lovelace

iantrich committed:
address comments

iantrich commented:
some updates from my testing:

* `action: navigation` correctly navigates to the proper view and the dialog remains displayed
* `action: card` inside an `action: card` swaps out the dialog

I think this is exactly how I want this to work.

@thomasloven as the resident popup expert, anything else you think should be tested or added/changed? Thanks.

thomasloven commented:
I'll have to agree with Bram on this one.
If there's one thing I've learned from popup-card it's that this kind of thing is HARD to wrap your head around. And that's something lovelace shouldn't be.

Misiu commented:
@thomasloven what do You think about adding a title to popup? IF someone adds a title to the card in card action then I think it should be displayed.
@iantrich All standard properties of cards will be supported? For example from [entities card]?(https://www.home-assistant.io/lovelace/entities/)

bramkragten commented:
> @thomasloven what do You think about adding a title to popup? IF someone adds a title to the card in card action then I think it should be displayed.
> @iantrich All standard properties of cards will be supported? For example from [entities card]?(https://www.home-assistant.io/lovelace/entities/)

Yes, it just renders a card the same way as Lovelace would

Misiu commented:
@bramkragten so the title will be there (if I add it as card property).
BTW what happened to the netlify links for PR's? They were a great way to check how the change will look and behave in user scenarios.

bramkragten commented:
Yes.

They where removed because we ran out of build minutes on netlify.

Misiu commented:
@bramkragten thank You for confirmation 🙂

iantrich commented:
@bramkragten @thomasloven 
A "messy" config seems like a silly limitation especially as we move further along with UI editors. Stacks within stacks is just like this, after all.

bramkragten commented:
I've been thinking about this, and I think it belongs to the `more-info` instead of a separate action. I think it would be nicer if we could give a custom `more-info` dialog. And that ofc can also be a card.

iantrich committed:
address comments

bramkragten commented on src/panels/lovelace/common/handle-click.ts:
```diff
@@ -53,7 +54,9 @@ export const handleClick = (
     case "more-info":
-      if (config.entity || config.camera_image) {
+      if (actionConfig.card) {
+        showCardDialog(node, { card: actionConfig.card });
```
I was thinking of using the actual `more-info-controls` and overruling the `more-info-content` with a `type` like we do for cards and badges.

We could later also add options to hide things like history or state.

Misiu replied:
@bramkragten I think that history, state, and the default controls should be hidden by default, otherwise this won't look good.

bramkragten replied:
The default controls is `more-info-controls` and if you want to make a generic customisable `more-info`, the history and state should not be hidden by default.

I don't want this to be a `more-info` dialog for Lovelace cards, it should be a `more-info` dialog you can customize. Otherwise we will have all different kind of dialogs in the future and it would be a pain to maintain.

Secondly, if we show a more info, it should also behave as a more info.

iantrich replied:
That would make this dependent on an entity which is should not be

iantrich replied:
If someone wants to show history or state, they can do that with cards. Only thing missing would be attributes perhaps, an attributes card might be a nice addition though to Lovelace

iantrich committed:
Revert move to more-info action. A distinct action is better suited, IMO

iantrich commented:
I've reverted this to its own action type, as I don't think merging this with more-info makes much sense as that would make it dependent on an entity, which I don't think this action should be. I'm also split on if the action should be `card`, `popup` or `show-dialog`
